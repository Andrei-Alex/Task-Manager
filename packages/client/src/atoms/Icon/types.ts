import { CustomStyles } from "../../../../libs/sharedTypes/styles";

export type Icons =
  | "HiMail"
  | "MdOutlinePassword"
  | "RxHamburgerMenu"
  | "IoIosArrowDown"
  | undefined;
export interface IIcon {
  extraStyles?: CustomStyles;
  iconName: Icons;
  size?: number;
  color?: string;
}
