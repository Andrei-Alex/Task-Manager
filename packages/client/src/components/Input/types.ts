import { CustomStyles } from "../../../../libs/sharedTypes/styles";
import { UseFormRegisterReturn } from "react-hook-form";
import type { Icons } from "../../atoms/Icon";
export interface IInput {
  label: string;
  required?: boolean;
  width?: string;
  placeholder?: string;
  id: string;
  containerStyle?: CustomStyles;
  register?: UseFormRegisterReturn<string>;
  icon?: Icons;
}
