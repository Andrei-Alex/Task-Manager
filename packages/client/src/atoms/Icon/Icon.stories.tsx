import type { Meta, StoryObj } from "@storybook/react";
import React from "react";
import {Icon} from "./Icon";

const meta: Meta<typeof Icon> = {
    component: Icon,
    title: "Atoms/Icon",
    tags: ["autodocs"],
    argTypes: {
        iconName: {
            name: "Select icon",
            description: "Display Icon",
            control: {
                type: 'radio',
            },
        },
    },

};
export default meta;
type Story = StoryObj<typeof Icon>;

export const Default: Story = {
    args: {
        extraStyles: {},
        iconName : "HiMail",
        color: "black",
        size: 16,
    },
};
