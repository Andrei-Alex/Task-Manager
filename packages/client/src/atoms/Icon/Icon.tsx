import React from "react";
import { HiMail } from "react-icons/hi";
import { MdOutlinePassword } from "react-icons/md";
import { RxHamburgerMenu } from "react-icons/rx";
import { IoIosArrowDown } from "react-icons/io";
import { Icons, IIcon } from ".";

/**
 * The Icon component is a versatile and reusable React component designed to render different
 * icons based on the specified iconName.
 * Leveraging the power of the react-icons library, it provides a simple and intuitive way to
 * incorporate icons into a React applications.
 *
 * With a straightforward usage, you can easily customize the appearance of the icons by adjusting
 * the color and size props.
 *
 * To enhance flexibility, the Icon component accepts an optional extraStyles prop
 * allowing you to apply additional styling to the icon container. This feature ensures seamless
 * integration into the application's design and layout.
 *
 * To optimize performance, the component is efficiently memoized using React.Memo along with a custom
 * comparison function. This ensures that unnecessary re-renders are minimized, resulting in a smoother
 * and more responsive user experience.
 *
 * ## Usage
 * ```JSX
 *    <Icon iconName="HiMail" color="blue" size={24} />
 * ```
 *
 * @component Icon
 * @param {Object} extraStyles - Additional styles to be applied to the icon container.
 * @param {string} iconName - The name of the icon to be rendered.
 * @param {string} color - The color of the icon. Defaults to "black".
 * @param {number} size - The size of the icon in pixels. Defaults to 16.
 * @returns {JSX.Element} - The rendered icon component.
 *
 */
export const Icon: React.FC<IIcon> = ({
  extraStyles,
  iconName,
  color = "black",
  size = 16,
}) => {
  const icon = (() => {
    switch (iconName as Icons) {
      case "HiMail":
        return <HiMail size={size} color={color} data-testid="HiMail-icon" />;
      case "MdOutlinePassword":
        return (
          <MdOutlinePassword
            size={size}
            color={color}
            data-testid="MdOutlinePassword-icon"
          />
        );
      case "RxHamburgerMenu":
        return (
          <RxHamburgerMenu
            size={size}
            color={color}
            data-testid="RxHamburgerMenu-icon"
          />
        );
      case "IoIosArrowDown":
        return (
          <IoIosArrowDown
            size={size}
            color={color}
            data-testid="IoIosArrowDown-icon"
          />
        );

      default:
        return null;
    }
  })();

  return <div style={{ height: `${size}px`, ...extraStyles }}>{icon}</div>;
};

function PropsAreEqual(prevProps: IIcon, nextProps: IIcon) {
  return (
    JSON.stringify(prevProps.extraStyles) ===
      JSON.stringify(nextProps.extraStyles) &&
    prevProps.iconName === nextProps.iconName &&
    prevProps.color === nextProps.color &&
    prevProps.size === nextProps.size
  );
}

export default React.memo(Icon, PropsAreEqual);
