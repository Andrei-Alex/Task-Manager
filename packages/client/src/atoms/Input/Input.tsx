import React from "react";
import { styles, IInput } from ".";

/**
 * Simple Input.
 *
 * ## Usage
 * ```jsx
 * <Label>{Text *}</Label>
 * <Input />
 * ```
 * @param {string} label - Text label
 * @param {boolean} required - If true add *
 * @param {number} width - If true add *
 * @param {number} placeholder - Add placeholder text
 * @param {number} id - Add id (label and htmlFor)
 * @param {objectr} containerStyle - style object
 * @return {JSX} Display Label with conditional * and Input
 */
export const Input: React.FC<IInput> = ({
  label,
  required = false,
  width = 415,
  placeholder = "Ex. Placeholder",
  id,
  containerStyle,
}) => {
  return (
    <div
      className={styles.container}
      style={{ width: `${width}px`, ...containerStyle }}
    >
      <div className={styles.labelContainer}>
        <label className={styles.label} htmlFor={id}>
          {label}
          {required && " *"}
        </label>
      </div>
      <input className={styles.input} placeholder={placeholder} id={id} />
    </div>
  );
};

export default Input;
