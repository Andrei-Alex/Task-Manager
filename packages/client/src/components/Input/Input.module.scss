@import '../../styles/variables';
@import '../../styles/mixin';
.container {
  display: flex;
  flex-direction: column;
  font-family: $fontPrimary;
}
.labelContainer{
 display: flex;
  margin-bottom: 10px;
  height: 25px;
}
.label {
  font-size: 22px;
  @include gradientText();
  color: $darkGray;
}
.inputContainer {
  position: relative;
}
.input {
  background-color: white;
  color: $darkGray;
  border: none;
  outline: none;
  border-bottom: 1px solid rgba(205,0,185,1);
  padding-left: 35px;
  padding-bottom: 10px;
  font-family: $fontPrimary;
  font-weight: 700;
  font-size: 16px;
  &:focus {
    outline-color: transparent;
  }
}
.iconContainer {
  position: absolute;
  z-index: 1;
}
.alert {
  transform: translateY(-15px);
  color: red;
}
