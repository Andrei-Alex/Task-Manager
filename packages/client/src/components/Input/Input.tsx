import React from "react";
import { styles, IInput } from "./index";
import { Icon } from "../../atoms";

/**
 * Simple Input.
 *
 * ## Usage
 * ```jsx
 * <Label>{Text *}</Label>
 * <Input />
 * ```
 * @param {string} icon Icon name
 * @param {string} label Text label
 * @param {boolean} required If true add *
 * @param {number} width If true add *
 * @param {number} placeholder Add placeholder text
 * @param {number} id Add id (label and htmlFor)
 * @param {object} containerStyle style object
 * @param {object} register react-hook-form object
 * @param {function} onChangeHandler onChange handler
 * @return {JSX} Display Label with conditional * and Input
 */
export const Input: React.FC<IInput> = ({
  icon,
  label,
  required = false,
  width = "100%",
  placeholder = "Ex. Placeholder",
  id,
  containerStyle,
  register,
  onChangeHandler,
}) => {
  return (
    <div className={styles.container} style={{ ...containerStyle }}>
      <div className={styles.labelContainer}>
        <label className={styles.label} htmlFor={id}>
          {label}
        </label>
        <p className={styles.alert}> {required && " *"}</p>
      </div>
      <div className={styles.inputContainer}>
        {icon ? (
          <i className={styles.iconContainer}>
            <Icon iconName={icon} color={"rgba(205,0,185,1)"} size={22} />
          </i>
        ) : null}
        <input
          onKeyDown={() => {
            if (onChangeHandler) {
              onChangeHandler();
            }
          }}
          className={styles.input}
          style={{ width: width }}
          placeholder={placeholder}
          id={id}
          {...register}
        />
      </div>
    </div>
  );
};

export default Input;
