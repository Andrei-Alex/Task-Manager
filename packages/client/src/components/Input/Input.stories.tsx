import type { Meta, StoryObj } from "@storybook/react";
import React from "react";
import {Input} from "./Input";

const meta: Meta<typeof Input> = {
  component: Input,
  title: "Components/Input",
  tags: ["autodocs"],
  argTypes: {
    label: {
      name: "Label",
      type: {
        name: "string",
        required: true,
      },
      defaultValue: "Default input",
      description: "Display text",
      control: {
        type: "text",
      },
    },
    required: {
      name: "Required",
      type: {
        name: "boolean",
        required: false,
      },
      defaultValue: false,
      description: "If true add *",
      control: {
        type: "boolean",
      },
    },
    width: {
      name: "Width value ",
      type: {
        name: "number",
        required: false,
      },
      defaultValue: 415,
      description: "Set container width",
      control: {
        range: true,
        min: 450,
        max: 1000,
      },
    },
    placeholder: {
      name: "Placeholder",
      type: {
        name: "string",
        required: false,
      },
      defaultValue: "Default input",
      description: "Set placeholder",
      control: {
        type: "text",
      },
    },
    id: {
      name: "id",
      type: {
        name: "string",
        required: true,
      },
      description: "Input ID",
      control: {
        type: "text",
      },
    },
  },
};
export default meta;
type Story = StoryObj<typeof Input>;

export const Default: Story = {
  args: {
    label: "Label",
    required: false,
    icon: "HiMail",
  },
};
