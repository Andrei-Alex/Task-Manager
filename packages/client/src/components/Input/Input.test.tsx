import React from "react";
import { fireEvent, render } from "@testing-library/react";
import { screen } from "@testing-library/dom";
import Input from "./Input";
import renderer from "react-test-renderer";
import Icon from "../../atoms/Icon/Icon";
const setup = () => {
  const renderInput = render(
    <Input label={"Label"} id={"label"} placeholder={"My Placeholder"} />
  );
  const input: HTMLInputElement = screen.getByLabelText("Label");
  return {
    input,
    ...renderInput,
  };
};

describe("Input", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <Input label={"Label"} id={"label"} placeholder={"My Placeholder"} />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("should have value", () => {
    const { input } = setup();
    fireEvent.change(input, { target: { value: "Jest" } });
    expect(input.value).toBe("Jest");
  });
  it("should display placeholder and label", () => {
    setup();
    const label = screen.getByText("Label");
    const placeholder = screen.getByPlaceholderText("My Placeholder");
    expect(label).toBeInTheDocument();
    expect(placeholder).toBeInTheDocument();
  });
  it("should display *", () => {
    render(<Input label={"Required label"} id={"Required"} required={true} />);
    const requiredSymbol = screen.getByText("*");
    expect(requiredSymbol).toBeInTheDocument();
  });
});
