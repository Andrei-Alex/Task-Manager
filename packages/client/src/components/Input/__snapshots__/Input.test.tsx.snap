// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders correctly 1`] = `
<div
  className="container"
  style={Object {}}
>
  <div
    className="labelContainer"
  >
    <label
      className="label"
      htmlFor="label"
    >
      Label
    </label>
    <p
      className="alert"
    >
       
    </p>
  </div>
  <div
    className="inputContainer"
  >
    <input
      className="input"
      id="label"
      placeholder="My Placeholder"
      style={
        Object {
          "width": "100%",
        }
      }
    />
  </div>
</div>
`;
